File: go/luhn/luhn.go
package luhn

func Valid(input string) bool {
	var n, digit, sum int
	for i := len(input) - 1; i >= 0; i-- {
		r := input[i]
		switch {
		case r == ' ':
			continue
		case r >= '0' && r <= '9':
			n++
			digit = int(r - '0')
			if n%2 == 0 {
				digit *= 2
			}
			if digit > 9 {
				digit -= 9
			}
			sum += digit
		default:
			return false
		}
	}
	return n > 1 && sum%10 == 0
}

File: lua/luhn/luhn.lua
local luhn = {}

function luhn.valid(number_text)
    local nows = number_text.gsub(number_text, " ", "")

    if #nows <= 1 or nows:find("%D") then
        return false
    else
        local sum = 0
        for index = #nows, 1, -1 do
            local digit = tonumber(nows:sub(index, index))
            if index % 2 == 0 then
                digit = digit * 2
                if digit > 9 then
                    digit = digit - 9
                end
            end
            sum = sum + digit
        end
        return sum % 10 == 0
    end
end

return luhn
